packages/next-server/lib/head.tsx

```diff
@@ -137,32 +137,60 @@ function reduceComponents(
     })
 }
 
-const Effect = withSideEffect()
-
 /**
  * This component injects elements to `<head>` of your page.
  * To avoid duplicated `tags` in `<head>` you can use the `key` property, which will make sure every tag is only rendered once.
  */
-function Head({ children }: { children: React.ReactNode }) {
-  return (
-    <AmpStateContext.Consumer>
-      {ampState => (
-        <HeadManagerContext.Consumer>
-          {updateHead => (
-            <Effect
-              reduceComponentsToState={reduceComponents}
-              handleStateChange={updateHead}
-              inAmpMode={isInAmpMode(ampState)}
-            >
-              {children}
-            </Effect>
-          )}
-        </HeadManagerContext.Consumer>
-      )}
-    </AmpStateContext.Consumer>
-  )
+function Head({ children }: { children: JSX.Element }) {
+  const ampState = React.useContext(AmpStateContext)
+  const updateHead = React.useContext(HeadManagerContext)
+  const instanceRef = React.useRef(children)
+
+  // Update the instanceRef every render
+  instanceRef.current = children
+
+  const inAmpMode = isInAmpMode(ampState)
+
+  const emitUpdate = React.useCallback(() => {
+    headState = reduceHeadInstances([...headInstanceRefs], {
+      inAmpMode,
+    })
+    if (updateHead) {
+      updateHead(headState)
+    }
+  }, [updateHead, reduceHeadInstances, inAmpMode])
+
+  // Since this is a Set(), we can just always add it
+  // We need to do ensure that we do this before the component
+  // renders for the first time anyway. So bing bang boom!
+  headInstanceRefs.add(instanceRef)
```

Comment:
This needs to be in a on-mount-only effect.

```suggestion
  React.useEffect(() => {
    headInstanceRefs.add(instanceRef)
  }, [])
```

So it should probably be merged with the effect below and then re-ordered.